### PR TITLE
rfc: add prelu post operation

### PR DESCRIPTION
**Pull request by piotrchmiel**
RFC for adding prelu post operation to oneDNN. [Rendered document.](https://github.com/piotrchmiel/oneDNN/blob/pchmiel/prelu_rfc/rfcs/20210518-prelu-postoperation/README.md)

### CODE REVIEW

File: rfcs/20210518-prelu-postoperation/README.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+![](graph_prelu.png "Graph attention layer")
+
+It becomes a bottleneck for the Attention layer in DIEN. Currently oneDNN
+supports PReLu only as
+[standalone primitive](https://oneapi-src.github.io/oneDNN/dev_guide_prelu.html).
+
+## Proposal
+
+Extend [supported postops list](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_post_ops.html)
+by adding PReLU forward postop.
```
Comment (kwiersch) [inline rfcs/20210518-prelu-postoperation/README.md:17]: I assume this is only for forward inference, but it should still be mentioned

File: rfcs/20210518-prelu-postoperation/README.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    void get_params_prelu(memory::desc &weights_desc) const {
+        const dnnl_memory_desc_t *data;
+        error::wrap_c_api(
+                dnnl_post_ops_get_params_prelu(get(), index, &c_alg, &data),
+                "could not get parameters of a prelu post-op");
+        weights_desc.data = *data;
+    }
+
+/// Arguments for fused prelu forward.
+#define DNNL_ARG_ATTR_POST_PRELU_WEIGHTS 10240
```
Comment (kwiersch) [inline rfcs/20210518-prelu-postoperation/README.md:50]: `10240 = 8192 + 2048` so this is the same as `DNNL_ARG_ATTR_POST_OP_DW | DNNL_ARG_MULTIPLE_DST`, which doesn't make sense to me. actually, i don't think we need to introduce any new argument, but can use `DNNL_ARG_ATTR_MULTIPLE_POST_OP(index) | DNNL_ARG_WEIGHTS` instead (as mentioned in the open questions section below).

File: rfcs/20210518-prelu-postoperation/README.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+```
+
+Assumptions:
+- Only one prelu post operation is allowed per primitive;
```
Comment (kwiersch) [inline rfcs/20210518-prelu-postoperation/README.md:54]: is there any fundamental technical reason for this limitation? if not, it might make sense to mention that this restriction can be lifted later on if needed

File: rfcs/20210518-prelu-postoperation/README.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                "could not append an prelu forward post-op");
+    }
+
+    /// Returns the parameters of a prelu forward post-op.
+    ///
+    /// @param weights_desc Output memory descriptor of a second operand.
+    void get_params_prelu(memory::desc &weights_desc) const {
+        const dnnl_memory_desc_t *data;
+        error::wrap_c_api(
+                dnnl_post_ops_get_params_prelu(get(), index, &c_alg, &data),
```
Comment (igorsafo) [inline rfcs/20210518-prelu-postoperation/README.md:44]:
```suggestion
                dnnl_post_ops_get_params_prelu(get(), index, &data),
```

File: rfcs/20210518-prelu-postoperation/README.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+/// Arguments for fused prelu forward.
+#define DNNL_ARG_ATTR_POST_PRELU_WEIGHTS 10240
+```
+
+Assumptions:
+- Only one prelu post operation is allowed per primitive;
+- Weights broadcasts semantics remains the same as in the standalone version.
+
+## Open questions
+1. Can we use mechanism from binary postops DNNL_ARG_ATTR_MULTIPLE_POST_OP in
```
Comment (igorsafo) [inline rfcs/20210518-prelu-postoperation/README.md:58]: Yes, `DNNL_ARG_ATTR_MULTIPLE_POST_OP` will simplify both API and implementation of prelu post op, because it will be aligned with others in chain of post ops.

File: rfcs/20210518-prelu-postoperation/README.md (hunk 1 above)
Comment (piotrchmiel) [reply rfcs/20210518-prelu-postoperation/README.md:17]: Updated.

File: rfcs/20210518-prelu-postoperation/README.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    /// The post-op can be defined as:
+    ///
+    ///      dst[:] <- prelu(dst[:], weights[:])
+    ///      prelu:
+    ///      dst[:] <- dst[:] if dst[:] > 0
+    ///      dst[:] <- dst[:] * weights[:] if dst[:] <= 0
+    ///
+    ///
+    /// @param weights_desc Learnable parameter memory descriptor.
+    void append_prelu_forward(const memory::desc &weights_desc) {
```
Comment (igorsafo) [inline rfcs/20210518-prelu-postoperation/README.md:33]: Do we allow to pass a particular memory tag or it should be any?

File: rfcs/20210518-prelu-postoperation/README.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
+# Introducing PReLU post operation support in oneDNN
+
+## Motivation
+
+This paper https://arxiv.org/abs/1502.01852 introduced PReLU activation
+function. The DIEN model uses PReLu as matmul activation.
+This gets expanded in TF graph.
+![](graph_prelu.png "Graph attention layer")
+
+It becomes a bottleneck for the Attention layer in DIEN. Currently oneDNN
```
Comment (igorsafo) [inline rfcs/20210518-prelu-postoperation/README.md:10]: What is expected performance improvement?

Comment (piotrchmiel) [reply rfcs/20210518-prelu-postoperation/README.md:10]: For know I don't have exact values. I'm trying for several days reach requestor and get answers about more details. I will go back to you as soon as possible.

File: rfcs/20210518-prelu-postoperation/README.md (hunk 6 above)
Comment (piotrchmiel) [reply rfcs/20210518-prelu-postoperation/README.md:33]: I think that the behaviour should follow standalone version - where we support "any tag" and encourage to use it. In case of jitted version, perfectly memory tag should be exactly the same as memory tag of destination tensor of the primitive it is being fused to. On the other hand in case of binary postops as far as I know we don't support "any" tag. In my opinion in case of prelu w should support it. What do you think ?

Comment (mgouicem) [reply rfcs/20210518-prelu-postoperation/README.md:33]: Sorry for digging up a thread that is 1 month old. Here I am afraid that using mem::desc here would expose too many possible "unsupported" usages by the user. Hence my question is why not use something similar to output scale (so a mask and weights of shape x)? This is more restrictive, but should capture exactly the use cases for prelu (typically, weights are per-channel I believe).

File: rfcs/20210518-prelu-postoperation/README.md (hunk 7 above)
Comment (piotrchmiel) [reply rfcs/20210518-prelu-postoperation/README.md:10]: Framework team is optimizing  FSRCNN on GPU from a customer. The network is mainly composed of conv + prelu pattern.

Now: 8.7ms(conv ) + 4.5ms(prelu, not fused) = 13.2ms
Expectation: 8.7ms(conv) + 0.4ms(10x perf gain by fusing) = 9.1ms

File: rfcs/20210518-prelu-postoperation/README.md (hunk 6 above)
Comment (piotrchmiel) [reply rfcs/20210518-prelu-postoperation/README.md:33]: No problem. I was still waiting for feedback from fwk team. This approach is consistent with passing tensor in binary postops. Why do you think that there will be many unsupported cases ? Currently reference prelu kernel is supporting all types of weights broadcast. I think that we can reuse existing code in reference kernels, so there won't be any unsupported situations. Would I be able to use the "any" tag in your proposal? In case jit kernels for example convolution weights should have exactly the same format as output (case where user requests per tensor broadcast).

File: rfcs/20210518-prelu-postoperation/README.md (hunk 7 above)
Comment (piotrchmiel) [reply rfcs/20210518-prelu-postoperation/README.md:10]: Generally expected performance improvement should be similar to fusing binary postops, here main cost will be loading weights, main advantage will be the presence prelu src tensor in registers.

File: rfcs/20210518-prelu-postoperation/README.md (hunk 6 above)
Comment (mgouicem) [reply rfcs/20210518-prelu-postoperation/README.md:33]: The main difference is:
- the mask approach assumes that memory is in `x` layout, so there is no special path to support blocked layout, padding, ...
- the memory descriptor approach can describe a larger variety of layouts, and it is generally hard to support all layouts + broadcast combination in jit injector.

In general, unless we have a path to always dispatch the best implementation possible with the memory descriptor approach, I would advocate for the mask approach.

Comment (piotrchmiel) [reply rfcs/20210518-prelu-postoperation/README.md:33]: Let's consider example. If prelu is post operation to convolution. Requested type of broadcast is per tensor (full broadcast). User set "any" tag as convolution destination (finally it can be blocked, plain etc). What memory layout should have weights tensor in that case ? In my opinion memory layout of weights should be exactly the same as convolution destination to generate most efficient and easy jit code without gather operations etc. If I understand correctly in mask approach the memory layout of the weights would be defined in advance e.g nchw ? Correct ?

File: rfcs/20210518-prelu-postoperation/README.md (hunk 7 above)
Comment (piotrchmiel) [reply rfcs/20210518-prelu-postoperation/README.md:10]: @igorsafo Do you need more details or can the the thread be resolved?

File: rfcs/20210518-prelu-postoperation/README.md (hunk 6 above)
Comment (mgouicem) [reply rfcs/20210518-prelu-postoperation/README.md:33]: This case is a bit different than binary postop. In binary postop, the second argument generally comes from another layer, so it indeed can have different layout and have those advanced broadcast semantics.
For the Prelu postop, the second argument is a weight for the layer. Also, keep in mind post-ops target inference as of now.

Now on the shape of weights. From the [paper](https://arxiv.org/pdf/1502.01852.pdf), the [pytorch definition](https://pytorch.org/docs/stable/generated/torch.nn.PReLU.html), and the few topologies I know of that use prelu, I believe the dominant use case is to have 1 negative slope per channel (so layout is 1d and `C`). So the rigidity of the mask approach will likely be transparent for performance.

On the other hand, if we go with the md approach:
- if we support `any` and return blocked layouts, this actually could hurt performance by triggering extra reorders.
- if the layout is not `any` and the injector does not support it, we would have to fallback to reference code, which completely defeats the purpose of fusion as `ref convolution + fused prelu` will likely be much slower than `jitted convolution -> prelu primitive`. So from a user standpoint, fusing with postops could be slower than not fusing for those cases.

So my point is, sometimes it is better to have a more rigid API with predictable behavior, than to have a more general API for which only a few cases behave as expected (in that case, fused to most efficient impl).

Comment (piotrchmiel) [reply rfcs/20210518-prelu-postoperation/README.md:33]: I agree that in case 1d and per oc broadcasts mask approach is better, because data format doesn't matter. But in my opinion it matters in other cases. Let's consider per tensor broadcast. Mask approach can possibly lead to performance degradation here. Let's go back to convolution example: let's say convolution output has nhwc format. Convolution before storing data has calculated ouptupt in vmm simd registers. Each of these registers contains a range of output channels. If data format of weights will be in that case nchw instead same as conv output it will have two side effects:

- instead of using offset to destination from convolution output we will in runtime calculate appropriate weights offset
- to apply postops on each register we will have to use gather operation, instead of reading continuous memory as if the weights layout was the same as the conv output layout

Comment (mgouicem) [reply rfcs/20210518-prelu-postoperation/README.md:33]: I agree with your points. Note that this suboptimal case would equally apply for both options though.
However, when using mask, it seems that there are fewer such cases, and less potential for extra reorders.

Also, on the importance of this per-tensor case: how often do we need per tensor broadcast for prelu weights? To the best of my knowledge, prelu weights are almost always per channel.
And if we believe that this fusion will always be slower than when using a separate reorder, we could even restrict the weights to be either of shape {1} or shape {c}.

File: rfcs/20210518-prelu-postoperation/README.md [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    /// @param weights_desc Output memory descriptor of a second operand.
+    void get_params_prelu(int &weights_mask) const {
+        const dnnl_memory_desc_t c_mask;
+        error::wrap_c_api(
+                dnnl_post_ops_get_params_prelu(get(), index, &c_mask),
+                "could not get parameters of a prelu post-op");
+        weights_mask = c_mask;
+    }
+
+    void set_output_scales(int mask, const std::vector<float> &scales) {
```
Comment (mgouicem) [inline rfcs/20210518-prelu-postoperation/README.md:88]:
```suggestion
    void set_prelu_weights(int mask, const std::vector<float> &scales) {
```

File: rfcs/20210518-prelu-postoperation/README.md [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        error::wrap_c_api(
+                dnnl_primitive_attr_set_output_scales(
+                        get(), (dnnl_dim_t)scales.size(), mask, scales.data()),
+                "could not set output scales primitive attribute");
+
```
Comment (mgouicem) [inline rfcs/20210518-prelu-postoperation/README.md:93]: Please rename scales -> prelu_weights

Comment (mgouicem) [inline rfcs/20210518-prelu-postoperation/README.md:93]:
```suggestion

```

File: rfcs/20210518-prelu-postoperation/README.md [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    ///    Prelu weights tensor is passed in runtime execution phase. Prelu
+    ///    weights tensor data type is implicitly assumed as f32 using plain
+    ///    layout (a, ab, acb, acdb, acdeb)
+
+    /// @param weights_mask Defines the correspondence between the output tensor
+    ///     dimensions and the prelu weights tensor. The set i-th bit indicates
+    ///     that a dedicated weights value is used for each index along that
+    ///     dimension. Set the mask to 0 to use a common weights valuee
+    ///     for the whole output tensor.
+    void append_prelu_forward(int weights_mask) {
```
Comment (mgouicem) [inline rfcs/20210518-prelu-postoperation/README.md:72]: nit: regarding naming, we did not put forward in the name for dw postop. I would suggest to do the same here to align naming.

File: rfcs/20210518-prelu-postoperation/README.md [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        error::wrap_c_api(dnnl_post_ops_append_prelu(get(), &weights_desc),
+                "could not append an prelu forward post-op");
+    }
+
+    /// Returns the parameters of a prelu forward post-op.
+    ///
+    /// @param weights_desc Output memory descriptor of a second operand.
+    void get_params_prelu(int &weights_mask) const {
```
Comment (mgouicem) [inline rfcs/20210518-prelu-postoperation/README.md:80]: this one need to have an index parameter (the index of the postop).

File: rfcs/20210518-prelu-postoperation/README.md [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    ///    weights tensor data type is implicitly assumed as f32 using plain
+    ///    layout (a, ab, acb, acdb, acdeb)
+
+    /// @param weights_mask Defines the correspondence between the output tensor
+    ///     dimensions and the prelu weights tensor. The set i-th bit indicates
+    ///     that a dedicated weights value is used for each index along that
+    ///     dimension. Set the mask to 0 to use a common weights valuee
+    ///     for the whole output tensor.
+    void append_prelu_forward(int weights_mask) {
+        error::wrap_c_api(dnnl_post_ops_append_prelu(get(), &weights_desc),
```
Comment (mgouicem) [inline rfcs/20210518-prelu-postoperation/README.md:73]:
```suggestion
        error::wrap_c_api(dnnl_post_ops_append_prelu(get(), weights_mask),
```

File: rfcs/20210518-prelu-postoperation/README.md [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    ///         oh = 14, ow = 14; // convolution output params
+    ///     // unique weights per output channel
+    ///     vector<float> weights = { ... };
+    ///     int oc_dim = 1; // mb_dim = 0, channel_dim = 1, height_dim = 2, ...
+    ///
+    ///     // construct a convolution descriptor
+    ///     dnnl::convolution::desc conv_d;
+    ///
+    ///     dnnl::primitive_attr attr;
+    ///     attr.append_prelu(attr, oc, 1 << oc_dim);
```
Comment (mgouicem) [inline rfcs/20210518-prelu-postoperation/README.md:45]:
```suggestion
    ///     attr.append_prelu(attr, 1 << oc_dim);
```

**Review by mgouicem — APPROVED** [review]
